execute/executetest/table.go

```diff
@@ -470,3 +488,287 @@ func MustCopyTable(tbl flux.Table) flux.Table {
+		tt.finish(false)
+	})
+	tt.run(t, "Retain", func(tt *tableTest) {
+		// Retain should allow the column reader to be used outside of Do.
+		tt.do(func(tbl flux.Table) error {
+			if tbl.Empty() {
+				return nil
+			}
+
+			var crs []flux.ColReader
+			if err := tbl.Do(func(cr flux.ColReader) error {
+				cr.Retain()
+				crs = append(crs, cr)
+				return nil
+			}); err != nil {
+				return err
+			}
+
+			// We are outside of the Do call. The column reader should
+			// still be valid so we should be able to read a value from it.
+			checkColReaders := func() {
+				for _, cr := range crs {
+					v := execute.ValueForRow(cr, 0, 0)
+					tt.t.Logf("first value for first column is %v", v)
+				}
+
+				// If there were multiple column readers returned for a single table,
+				// we need to check that they did not share a buffer.
+				if len(crs) > 1 {
+					for i := 0; i < len(crs)-1; i++ {
+						for j := i + 1; j < len(crs); j++ {
+							if colReadersEqual(crs[i], crs[j]) {
+								tt.t.Errorf("retained column reader is the same as another column reader (%d, %d) for table %v", i, j, tbl.Key())
+							}
+						}
+					}
+				}
+			}
+			checkColReaders()
+
+			// Call Done (which should have already happened by the call to Do)
+			// and ensure that the above check still succeeds.
+			tbl.Done()
+			checkColReaders()
+
+			for _, cr := range crs {
+				cr.Release()
+			}
+			return nil
+		})
+		tt.finish(true)
+	})
+	tt.run(t, "Len", func(tt *tableTest) {
+		tt.do(func(tbl flux.Table) error {
+			return tbl.Do(func(cr flux.ColReader) error {
+				want := cr.Len()
+				for i, n := 0, len(cr.Cols()); i < n; i++ {
+					got := func(cr flux.ColReader, i int) int {
+						switch cr.Cols()[i].Type {
+						case flux.TFloat:
+							return cr.Floats(i).Len()
+						case flux.TInt:
+							return cr.Ints(i).Len()
+						case flux.TUInt:
+							return cr.UInts(i).Len()
+						case flux.TString:
+							return cr.Strings(i).Len()
+						case flux.TBool:
+							return cr.Bools(i).Len()
+						case flux.TTime:
+							return cr.Times(i).Len()
+						default:
+							panic(fmt.Errorf("unexpected column type: %v", cr.Cols()[i].Type))
+						}
+					}(cr, i)
+					if got != want {
+						tt.t.Errorf("column %d does not have the wanted length (%v) -want/+got:\n\t- %d\n\t+ %d", i, tbl.Key(), want, got)
+					}
+				}
+				return nil
+			})
+		})
+		tt.finish(true)
+	})
+}
+
+func colReadersEqual(a, b flux.ColReader) bool {
+	if a.Len() != b.Len() {
+		return false
+	}
+
+	if len(a.Cols()) != len(b.Cols()) {
+		return false
+	}
+
+	for i, n := 0, len(a.Cols()); i < n; i++ {
+		if a.Cols()[i] != b.Cols()[i] {
+			return false
+		}
+
+		// Compare the memory address for the arrow buffer.
+		// The same buffer contents with different memory addresses
+		// are considered different.
+		switch a.Cols()[i].Type {
+		case flux.TFloat:
+			if a.Floats(i) != b.Floats(i) {
+				return false
+			}
+		case flux.TInt:
+			if a.Ints(i) != b.Ints(i) {
+				return false
```

Comment:
Don't you want to be sure that they don't share any buffer? In this  way, it is enough not to share a single buffer and share the others and the readers are not equal.

Reply:
I don't think that's necessary. I check earlier to see if the metadata is the same so we're certain that we're comparing the same columns.

This is mostly a check to make sure that any buffering mechanism for column readers, like the one influxdb employs, doesn't end up modifying the old column reader when it is retained since there's specific code exercising that.

Reply:
Mmmm, I still think that it should be better to check for every col reader to be different. Maybe I am missing something 😅